tripal/tripal
# Resolve discovered field name length problems

# Bug Fix

### Closes #1928

### Tripal Version: 4.x

## Description
Check for new fields fails if property name is too long.

Problems:
A value of 50 is used for truncation, but the actual maximum is 32.
With simple truncation, two fields might therefore get identical names, not good.

Here I get the actual maximum length from Drupal rather than hardcoding it.
In Tripal 3 we appended the cvterm_id to truncated field names to guarantee uniqueness.
There a fixed length of 20 characters was used, then underscore and then cvterm_id.
Here, I use the maximum length that can fit in 32 characters so that field names will be as long as possible.

If a cvterm_id value is not readily available, the PHP `uniqid()` function is used to generate the suffix for the field name.

## Testing?
I added an automated test for the truncation and even threw in a unicode character for good measure. This revealed that multibyte characters were being replaced with multiple underscores, fixed that with a `u` suffix on the regex.

For manual testing:
1. Create an organism with a property
```
sitedb=> INSERT INTO chado.organism (genus, species) VALUES ('Tripalus', 'databasica');
INSERT 0 1
sitedb=> INSERT INTO chado.organismprop (organism_id, type_id, value) VALUES
  ((SELECT organism_id FROM chado.organism WHERE species='databasica'),
  (SELECT cvterm_id FROM chado.cvterm WHERE name='mitochondrial_genetic_code_name'), 'test');
INSERT 0 1
```
2. Go to Tripal -> Page Structure
3. For "Organism" select "Manage Fields"
4. +Check for new fields. You should see this
![2024-07-16_fixed](https://github.com/user-attachments/assets/a7ae29ac-9769-4bdd-8057-22afc325ae2e)
5. Click "Add Fields" button. There should be no errors.
![2024-07-16_fixed2](https://github.com/user-attachments/assets/7ec7f684-091a-4fac-b300-b6ffdaf17810)
